File: client/scripts/hello.py
import grpc
import hello_pb2
import hello_pb2_grpc

def dotest():
    channel = grpc.insecure_channel('192.168.70.128:20000')
    stub = hello_pb2_grpc.TesterStub(channel)
    req = hello_pb2.HelloReq()

    # Simple RPC
    ## sync
    req.msg = 'simple rpc.'
    msg = stub.SayHello(req)
    print(msg.msg)
    ## async
    rep = stub.SayHello.future(req)
    msg = rep.result()
    print(msg.msg)

    # Server-side RPC
    req.msg = 'server-side rpc.'
    for msg in stub.SayHello2(req):
        print(msg.msg)

    def genIterReq(num):
        while num > 0:
            yield req
            num -= 1

    # Client-side RPC
    req.msg = 'client-side rpc.'
    ## sync
    msg = stub.SayHello3(genIterReq(3))
    print(msg.msg)
    ## async
    rep = stub.SayHello3.future(genIterReq(3))
    msg = rep.result()
    print(msg.msg)

    # Bidirectional RPC
    req.msg = 'bidirectional rpc.'
    for msg in stub.SayHello4(genIterReq(3)):
        print(msg.msg)

File: server/internal/service/hello/hello.go
package hello

import (
	"context"
	"io"
	"jjgame/internal/logger"
	"strconv"
)

type HelloService struct {
	UnimplementedTesterServer
}

// Simple RPC
func (s HelloService) SayHello(ctx context.Context, msg *HelloReq) (*HelloResp, error) {
	logger.INFO_MSG(msg.Msg)
	resp := HelloResp{
		Msg: "Reply " + msg.Msg,
	}
	return &resp, nil
}

// Server-side streaming RPC
func (s HelloService) SayHello2(msg *HelloReq, stream Tester_SayHello2Server) error {
	logger.INFO_MSG(msg.Msg)
	for i := 1; i < 3; i++ {
		resp := HelloResp{
			Msg: "Reply " + msg.Msg + strconv.Itoa(i),
		}
		if err := stream.Send(&resp); err != nil {
			return err
		}
	}
	return nil
}

// Client-side streaming RPC
func (s HelloService) SayHello3(stream Tester_SayHello3Server) error {
	for {
		msg, err := stream.Recv()
		if err == io.EOF {
			resp := HelloResp{
				Msg: "Reply client-side rpc.",
			}
			return stream.SendAndClose(&resp)
		} else if err != nil {
			return err
		}
		logger.INFO_MSG(msg.Msg)
	}
}

// Bidirectional streaming RPC
func (s HelloService) SayHello4(stream Tester_SayHello4Server) error {
	for {
		msg, err := stream.Recv()
		if err == io.EOF {
			return nil
		} else if err != nil {
			return err
		}

		logger.INFO_MSG(msg.Msg)
		for i := 1; i < 3; i++ {
			resp := HelloResp{
				Msg: "Reply " + msg.Msg + strconv.Itoa(i),
			}
			if err := stream.Send(&resp); err != nil {
				return err
			}
		}
	}
}
